File: 前缀和/974_和可被K整除的子数组/sol.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        dic = {}
        total = count = 0
        for num in nums:
            total += num
            modulus = total % k
            same += dic.get(modulus, 0)
            count+=same
            dic[modulus] = same=1
        return count

        
        

File: 前缀和/974_和可被K整除的子数组/sol.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysDivByK = function (nums, k) {
  let map = new Map();
  map.set(0, 1);
  let total = 0,
    count = 0;

  for (let item of nums) {
    total += item;
    let modulus = ((total % k) + k) % k;
    if (map.get(modulus)) {
      count += map.get(modulus);
    }
    map.set(modulus, map.get(modulus) ? map.get(modulus) + 1 : 1);
  }

  return count;
};

const nums = [-1,2,9],
  k = 2;
console.log(subarraysDivByK(nums, k));
